### DISCUSSION TITLE
Add Kafka to TDP

### DISCUSSION THREAD

**Discussion by Nuttymoon** [TOSIT-IO/TDP#27] [New components] [answered]
### Presentation of the project

> [Apache Kafka](https://kafka.apache.org) is an open-source distributed event streaming platform used by thousands of companies for high-performance data pipelines, streaming analytics, data integration, and mission-critical applications.

Kafka is the de facto standard for building scalable fault-tolerant messaging systems.

### Why should it be included in TDP?

- As a big data Apache project, Kafka is **very well integrated with many Hadoop projects** that are part of TDP (Hive, Spark, Ranger, etc.)
- Kafka brings **complementary functionalities** that are needed to build full-fledged data lakes

### Targeted version of the project

[Kafka 3.2.x](https://github.com/apache/kafka/tree/3.2) (latest version)

**Note:** Unfortunately Kafka Raft metadata mode (that allows running Kafka without ZooKeeper) is still not production-ready in this release. See [KRaft (aka KIP-500) mode Preview Release](https://github.com/apache/kafka/blob/3.2/config/kraft/README.md).

#### ZooKeeper

Kafka 3.x uses ZooKeeper version 3.6.3 (see [dependencies.gradle](https://github.com/apache/kafka/blob/87d6d78913d4559aad6fb4c2ea19a0b8449a9ab9/gradle/dependencies.gradle#L116)) whereas the version packaged in TDP is 3.4.6.

Here are a few things that I have noted:
  - **Building Kafka 3.2 using ZK 3.4.6 as a dependency fails**. Kafka 3.x uses new client APIs introduced in ZK 3.5.2 (e.g. [ZKClientConfig](https://github.com/apache/zookeeper/blob/master/zookeeper-server/src/main/java/org/apache/zookeeper/client/ZKClientConfig.java))
  - If we look at [Confluent interoperability documentation](https://docs.confluent.io/platform/current/installation/versions-interoperability.html#zk) we can see that ZK 3.4.6 can only be used with Kafka < 0.11, which is clearly not a viable option for TDP

This brings me to the following questions: Can we integrate another version of ZooKeeper into TDP? If yes, how will we modify [tdp-collection](https://github.com/TOSIT-IO/tdp-collection)'s architecture?

**Comment by leopaul36** [comment, marked as answer]
I think we will all agree that Kafka fits well into TDP due to its integration with the other components.

As for the ZooKeeper dependency, we can try and see if our current K̶a̶f̶k̶a̶ ZooKeeper role is able to deploy a ZooKeeper in version 3.6.x. If not maybe we can do a ZooKeeper subtask inside the Kafka role like we did for Solr in the Ranger role ?

What do you think @gboutry @rpignolet @nschung ?

**Reply by Nuttymoon** [reply]
@leopaul36 did you mean
> see if our current **~~Kafka~~ ZooKeeper** role is able to deploy a ZooKeeper in version 3.6.x

**Reply by leopaul36** [reply]
Yes, my bad.

**Reply by rpignolet** [reply]
Doing ZooKeeper subtask inside Kafka involves having 2 ZooKeeper clusters running at the same time ?

**Reply by leopaul36** [reply]
> Doing ZooKeeper subtask inside Kafka involves having 2 ZooKeeper clusters running at the same time ?

Yes, it is always a good practice to have a dedicated ZooKeeper cluster for Kafka anyway.

**Reply by Nuttymoon** [reply]
Ok I will try reusing the current ZooKeeper role and create a new independent service `zookeeperkafka`. It should work fine.

**Comment by nschung** [comment]
Kafka should be integrated as an option. It can be deployed in dedicated architecture. Therefore, Kafka can use its compatible ZK version IMO.
However, it's better to have the same ZK version of TDP and Kafka.

**Comment by rpignolet** [comment]
When reviewing PR related to Kafka, I think we didn't discuss enough about adding Kafka to TDP.

For me the goal, when adding a service/component to TDP, is to have a consistent ecosystem i.e. the new service/component should interact with other service/component inside TDP.

Another point is about the dedicated ZooKeeper, we should talk about it because, just to add Kafka we deploy another ZooKeeper cluster and this is really the good solution ?

I think that it should, at least, be moved to `tdp-collection-extras` and not in `tdp-collection` core if we absolutely want to have it.

**Comment by Edouard-R** [comment]
As suggested in Issue #340 of tdp-collection, starting with a Kafka 2.8.1 may be a good idea.
If we can't find a recent version of Kafka (2.X) compatible with ZK 3.4.6, maybe we should simply wait for the moment we'll fix an entire new set of versions for all our component.

**Comment by Nuttymoon** [comment]
As suggested by @rpignolet and @Edouard-R I am going to move Kafka and it's dedicated ZooKeeper to `tdp-collection-extras` for now. We can reconsider a graduation to `tdp-collection` later on.
